components/script/stylesheet_loader.rs

```diff
@@ -375,9 +496,268 @@ impl<'a> StyleStylesheetLoader for StylesheetLoader<'a> {
+/// needs to be done on the event loop, thus when the thread parsing the sheet
+/// is done it enqueues this task on the script thread's task queue that's
+/// managing the document.
+///
+/// This task is used for both top-level as well as nested stylesheet loads. For
+/// top-level sheets the `top_level_stylesheet` field must be set, but parsing
+/// an imported sheet returns an empty sheet wrapped in an Arc in an
+/// `ImportRule` (which is inserted in its parent's rule list), and after an
+/// async fetch and parse, the empty stylesheet is set through the Arc
+/// reference.
+struct FinishAsyncStylesheetLoadTask {
+    /// Used to tell the document whether the load has concluded successfully.
+    successful: bool,
+    /// Whether document's current stylesheet needs to be invalidated. This is
+    /// necessary if a new top-level stylesheet or an import sheet has been
+    /// successfully parsed.
+    invalidate_stylesheets: bool,
+    origin_clean: bool,
+    /// If this was a top-level stylesheet load, the parsed stylesheet is placed
+    /// here.
+    top_level_stylesheet: Option<Arc<Stylesheet>>,
+    /// The element for which the stylesheet was loaded.
+    elem: Trusted<HTMLElement>,
+    /// The document for which the stylesheet was loaded.
+    document: Trusted<Document>,
+    /// The URL of the stylesheet.
+    url: ServoUrl,
+    /// A token which must match the generation id of the `HTMLLinkElement` for
+    /// it to load the stylesheet. This is ignored for `HTMLStyleElement` and
+    /// imports.
+    request_generation_id: Option<RequestGenerationId>,
+}
+
+impl TaskOnce for FinishAsyncStylesheetLoadTask {
+    /// Set the parsed stylesheet (if any) on the link element if this was
+    /// a load for a top-level stylesheet. If this was the last missing
+    /// stylesheet for the document, fire a load event.
+    fn run_once(self) {
+        let elem = self.elem.root();
+        let document = self.document.root();
+
+        if let Some(sheet) = self.top_level_stylesheet {
+            let link = elem.downcast::<HTMLLinkElement>().unwrap();
+            // We must first check whether the generations of the context and
+            // the element match up, else we risk applying the wrong stylesheet
+            // when responses come out-of-order.
+            let is_stylesheet_load_applicable = self
+                .request_generation_id
+                .map_or(true, |gen| gen == link.get_request_generation_id());
+            if is_stylesheet_load_applicable {
+                if link.is_alternate() {
+                    sheet.set_disabled(true);
+                }
+                link.set_stylesheet(sheet);
+            }
+        }
+
+        // We also need to invalidate a document's stylesheets if this was an
+        // import rule load, not just in the case of top-level sheet loads.
+        if self.invalidate_stylesheets {
+            document.invalidate_stylesheets();
```

Comment:
Unless this is the issue you're running into with that test, in which case we'll need to fix it probably! Or we could mark it as failing for now pointing to this bug.

Reply:
I haven't run into issues since doing the changes described earlier, but I won't lie I might be a little too tired right now to understand what the pre-existing problem is, so I may be misunderstanding what you meant :) Why invalidate previous sheets and set the new stylesheet if there was no stylesheet that could be parsed?

Reply:
No, it's alright, let's not add behavior changes, and fix that in a followup if it's a problem.